File: app/src/main/java/com/seasonfif/learnkotlin/concurrent/ProducerConsumer.java
package com.seasonfif.learnkotlin.concurrent;

import java.util.LinkedList;
import java.util.Queue;
import java.util.Random;

public class ProducerConsumer {

    public static void main(String...args){
        Queue<Integer> queue = new LinkedList<>();

        new Produc(queue).start();
        new Consum(queue).start();
        new Consum(queue).start();
        new Consum(queue).start();
    }


    private static class Produc extends Thread{
        private final Queue<Integer> queue;

        public Produc(Queue<Integer> queue) {
            this.queue = queue;
        }

        @Override
        public void run() {
            while(true){
                synchronized (queue){
                    try {
                        sleep(1000);
                    } catch (InterruptedException e) {
                        e.printStackTrace();
                    }

                    if (queue.size() >= 10){
                        try {
                            queue.wait();
                        } catch (InterruptedException e) {
                            e.printStackTrace();
                        }
                    }

                    int num = new Random().nextInt(100) + 1;
                    queue.offer(num);

                    System.out.println(getName() + "生产：" + num);

                    queue.notifyAll();
                }
            }

        }
    }


    private static class Consum extends Thread{

        private final Queue<Integer> queue;

        public Consum(Queue<Integer> queue) {
            this.queue = queue;
        }

        @Override
        public void run() {
            while (true){
                synchronized (queue){
                    try {
                        sleep(1000);
                    } catch (InterruptedException e) {
                        e.printStackTrace();
                    }

                    if (queue.size() == 0){
                        try {
                            queue.wait();
                        } catch (InterruptedException e) {
                            e.printStackTrace();
                        }
                    }

                    int num = queue.poll();
                    System.out.println(getName() + "消费：" + num);

                    queue.notifyAll();

                }

            }

        }
    }
}

File: app/src/main/java/com/seasonfif/learnkotlin/concurrent/Producer-Consumer.kt
package com.seasonfif.learnkotlin.concurrent

import java.util.*

fun main(args: Array<String>) {

    val lock = Object()
    var queue : Queue<Int> = LinkedList()

    Producer(queue, 10, lock).start()
    Producer(queue, 10, lock).start()
    Consumer(queue, 10, lock).start()
    Consumer(queue, 10, lock).start()
    Consumer(queue, 10, lock).start()
}

class Producer : Thread{

    private var lock : Object
    private var queue : Queue<Int>
    private var maxSize : Int

    constructor(queue : Queue<Int>, maxSize : Int, lock : Object) : super(){
        this.queue = queue
        this.maxSize = maxSize
        this.lock = lock
    }

    override fun run() {
        synchronized(lock){
            while (true){
                sleep(1000)

                if (queue.size == maxSize){
                    lock.wait()
                }

                var num : Int = (Math.random() * 100).toInt()

                queue.offer(num)
                println(name + "生产：$num")

                lock.notifyAll()
            }
        }

    }
}

class Consumer : Thread{
    private var lock : Object
    private var queue : Queue<Int>
    private var maxSize : Int

    constructor(queue : Queue<Int>, maxSize : Int, lock : Object) : super(){
        this.queue = queue
        this.maxSize = maxSize
        this.lock = lock
    }

    override fun run() {
        while (true){
            synchronized(lock){
                sleep(1000)
                if (queue.isEmpty()){
                    lock.wait()
                }

                var num = queue.poll()
                println(name + "消费：$num")

                lock.notifyAll()
            }
        }
    }
}